File: run_reconcile_quick.js
const { Client } = require('typesense');
const https = require('https');

const TS_HOST = 'uibn03zvateqwdx2p-1.a1.typesense.net';
const TS_KEY = 'MPphr9zDlLzHRFQHDH4AyQb5hw2ugew7';
const NEON_CONN = 'postgresql://neondb_owner:npg_BvF2JsQ8drba@ep-calm-fire-ad0dfnqd-pooler.c-2.us-east-1.aws.neon.tech/wildphotography?sslmode=require';

const client = new Client({
  nodes: [{ host: TS_HOST, port: 443, protocol: 'https' }],
  apiKey: TS_KEY
});

async function run() {
  // Get Typesense count
  const tsResult = await client.collections('photos').documents().search({ q: '*', limit: 0 });
  const tsCount = tsResult.found;
  console.log('Typesense docs:', tsCount);

  // Get DB eligible count via raw query
  const { Pool } = require('pg');
  const pool = new Pool({ connectionString: NEON_CONN });
  const dbResult = await pool.query("SELECT COUNT(*) FROM photos WHERE search_ready = true AND status NOT IN ('archived', 'legacy_static')");
  const dbEligible = parseInt(dbResult.rows[0].count);
  console.log('DB eligible:', dbEligible);

  const drift = tsCount - dbEligible;
  console.log('Drift:', drift);

  if (Math.abs(drift) > 10) {
    console.log('Running reconciliation...');
    // Get eligible IDs
    const eligibleResult = await pool.query("SELECT id FROM photos WHERE search_ready = true AND status NOT IN ('archived', 'legacy_static')");
    const eligibleIds = new Set(eligibleResult.rows.map(r => String(r.id)));
    console.log('Eligible IDs count:', eligibleIds.size);

    // Export Typesense docs
    const exportRes = await client.collections('photos').documents().export();
    const lines = exportRes.split('\n').filter(l => l.trim());
    console.log('Typesense exported lines:', lines.length);

    let toDelete = 0;
    let deleted = 0;
    for (const line of lines) {
      try {
        const doc = JSON.parse(line);
        const pid = String(doc.id);
        if (pid && !eligibleIds.has(pid)) {
          toDelete++;
          await client.collections('photos').documents(pid).delete();
          deleted++;
        }
      } catch(e) {}
    }
    console.log(`Stale: ${toDelete}, Removed: ${deleted}`);
  } else {
    console.log('Drift within tolerance, no reconcile needed');
  }

  await pool.end();
  console.log('Done');
}

run().catch(e => { console.error('Error:', e.message); process.exit(1); });

File: run_reconcile_quick.py
#!/usr/bin/env python3
"""Quick Typesense reconciliation run"""
import os, sys, json, time, requests, psycopg2

NEON_CONN = "postgresql://neondb_owner:npg_BvF2JsQ8drba@ep-calm-fire-ad0dfnqd-pooler.c-2.us-east-1.aws.neon.tech/wildphotography?sslmode=require"
TS_HOST = "uibn03zvateqwdx2p-1.a1.typesense.net"
TS_KEY = "MPphr9zDlLzHRFQHDH4AyQb5hw2ugew7"
TS_COLL = "photos"
BATCH = 200

def ts_url(path):
    return f"https://{TS_HOST}:443{path}"

def ts_headers():
    return {"X-Typesense-Api-Key": TS_KEY, "Content-Type": "application/json"}

conn = psycopg2.connect(NEON_CONN)
cur = conn.cursor()

# Get Typesense current count
r = requests.get(ts_url(f"/collections/{TS_COLL}/documents/search"), headers=ts_headers(), params={"q": "*", "limit": 0}, timeout=15)
ts_count = r.json()["found"]
print(f"Typesense docs: {ts_count}")

# Get DB eligible count
cur.execute("SELECT COUNT(*) FROM photos WHERE search_ready = true AND status NOT IN ('archived', 'legacy_static')")
db_eligible = cur.fetchone()[0]
print(f"DB eligible: {db_eligible}")

drift = ts_count - db_eligible
print(f"Drift: {drift}")

# Reconcile: delete stale from Typesense where photo_id not in eligible set
if abs(drift) > 10:
    print("Running reconciliation...")
    cur.execute("SELECT id FROM photos WHERE search_ready = true AND status NOT IN ('archived', 'legacy_static')")
    eligible_ids = set(str(row[0]) for row in cur.fetchall())
    
    r = requests.get(ts_url(f"/collections/{TS_COLL}/documents/export"), headers=ts_headers(), timeout=60)
    lines = r.text.strip().split('\n')
    to_delete = []
    for line in lines:
        try:
            doc = json.loads(line)
            pid = str(doc.get('id', ''))
            if pid and pid not in eligible_ids:
                to_delete.append(pid)
        except:
            pass
    
    print(f"Stale docs to remove from Typesense: {len(to_delete)}")
    deleted = 0
    for pid in to_delete[:500]:
        try:
            requests.delete(ts_url(f"/collections/{TS_COLL}/documents/{pid}"), headers=ts_headers(), timeout=5)
            deleted += 1
        except:
            pass
    print(f"Removed {deleted} stale docs")
else:
    print("Drift within tolerance, no reconcile needed")

cur.close()
conn.close()
print("Done")